File: 02_DevApplication_IOS_avecObjectif-C/11_MapViewAvecListePointsInteretsEtTrajets/09_MapView/TrajetViewController.m
//
//  TrajetViewController.m
//  09_MapView
//
//  Created by etudiant on 24/04/2017.
//  Copyright © 2017 etudiant. All rights reserved.
//

#import "TrajetViewController.h"
#import <MapKit/MapKit.h>
@interface TrajetViewController ()
@property (weak, nonatomic) IBOutlet MKMapView *maMap;

@end

@implementation TrajetViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    //rendre la map transparente
    
    self.maMap.alpha = 0;
    //demander explicitement l'affichage de la position de l'utilisateur
    //car par defaut, la map ne l'affiche pas
    self.maMap.showsUserLocation = YES;
    //Déclarer que ce controleur est le délégué de la map
    //qu'il implémente certaines méthodes du protocole MKMapViewDelegate
    self.maMap.delegate = self;
}
//implémenter la méthode du protocol MKMapVieDelegate qui est appelée
//quand la position de l'utilisateur change
-(void)mapView:(MKMapView *)mapView didUpdateUserLocation:(MKUserLocation *)userLocation
{
   //est-ce la première fois que la position de l'utilisateur a été récupérée?
    if(self.maMap.alpha == 0)
    {
        //définir un zoom initial à l'aide d'une région de
        //disons 2000mx2000m centrée sur la positon de
        //l'utilisateur passée en paramère
        MKCoordinateRegion region = MKCoordinateRegionMakeWithDistance(CLLocationCoordinate2DMake(userLocation.coordinate.latitude, userLocation.coordinate.longitude), 2000,2000);
        
        //passer la région à la map
        self.maMap.region = region;
        //Rendre la map opaque
        self.maMap.alpha=1;
        
        //demande à un serveur de calculer le trajet
        [self calculerTrajet];
        
    }else
        self.maMap.centerCoordinate = userLocation.coordinate;
    
}
-(void) calculerTrajet
{
    // créer une demande de calcul de trajet
    MKDirectionsRequest* demandeTrajet = [[MKDirectionsRequest alloc] init];
    
    //renseigner le point de départ
    // dans mon cas, le point de départ est la position de l'utilisateur
    demandeTrajet.source = [MKMapItem ];
}
- (IBAction)retourEcranPrecedentTouched:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}
- (IBAction)bouttonRetourPremierEcranTouched:(id)sender {
}

- (IBAction)btnRecarculer:(id)sender {
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: 11_MapViewAvecListeEtTrajet/11_MapViewAvecListeEtTrajet/TrajetViewController.swift
//
//  TrajetViewController.swift
//  11_MapViewAvecListeEtTrajet
//
//  Created by etudiant on 5/30/17.
//  Copyright © 2017 etudiant. All rights reserved.
//

import UIKit
import MapKit
class TrajetViewController: UIViewController, MKMapViewDelegate {

    @IBAction func btnRetourTouched(_ sender: UIBarButtonItem) {
        dismiss(animated: true, completion: nil)
    }
    @IBOutlet weak var maMap: MKMapView!
    @IBAction func btnRecalculerTrajetTouched(_ sender: UIBarButtonItem) {
        calculerTrajet()
    }
    var destinationTrajet: MKMapItem?
    
    var tbInstructions : [String] = []
    override func viewDidLoad() {
        super.viewDidLoad()
        
        maMap.alpha = 0
        
        maMap.showsUserLocation = true
        maMap.delegate = self

        // Do any additional setup after loading the view.
    }
    
    func mapView(_ mapViex:MKMapView, didUpdate userLocation: MKUserLocation)
    {
        if(self.maMap.alpha == 0)
        {
            //définir un zoom initial à l'aide d'une région de
            //disons 2000mx2000m centrée sur la positon de
            //l'utilisateur passée en paramère
            let region = MKCoordinateRegionMakeWithDistance(CLLocationCoordinate2DMake(userLocation.coordinate.latitude, userLocation.coordinate.longitude), 2000,2000);
            
            //passer la région à la map
            self.maMap.region = region;
            //Rendre la map opaque
            self.maMap.alpha=1;
            
            //demande à un serveur de calculer le trajet
            maMap.alpha = 1
            calculerTrajet();
        }
            else
            {
                    maMap.centerCoordinate = userLocation.coordinate;
    
            }
    }
    
    func calculerTrajet()
    {
        //supprimer tous les trajts
        supprimerTousLesTrajets();
        
        //créer une demande de calcul
        let demandeTrajet = MKDirectionsRequest()
        
        //définir le point de départ
        //dans mon cas, le point de départ est la position de  l'utilisateur
        
        demandeTrajet.source = MKMapItem.forCurrentLocation()
        //definir le point d'arrivé à partir de l'écran précédent
        demandeTrajet.destination = destinationTrajet
        
        
        demandeTrajet.requestsAlternateRoutes = false
        let directions = MKDirections(request: demandeTrajet)
        
        directions.calculate{
            (argTrajet,argErreur) in
                if argErreur != nil
            {
            print("Le calcul du trajet a échoué, erreur invoquée:\(argErreur?.localizedDescription)")
            }
                else{
                    self.afficherTrajet(argTrajet!)
                    }
          
        }
    }

    func afficherTrajet(_ trajet: MKDirectionsResponse)
    {
        var nb = 1
        for route in trajet.routes{
            maMap.add(route.polyline, level: MKOverlayLevel.aboveRoads)
            //dessiner le trajet enrre la rue et l"étiquette contenant le nom de la rue 
            for etape in route.steps
            {
                //déposer les instructions pour chaque étape
                //dans le tabelau d'instructions
                tbInstructions.append("\(nb+=1),\(etape.instructions)")
             
            }
        }
    }
    
    //méthode du protocole
    func mapView(_ mapView:MKMapView, rendererFor argOverLay:MKOverlay)-> MKOverlayRenderer
    {
        let renduGraphique = MKPolylineRenderer(overlay: argOverLay)
        
        renduGraphique.strokeColor = UIColor.blue
        renduGraphique.lineWidth = 5.0
        
        return renduGraphique
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func supprimerTousLesTrajets()
    {
                for couche_a_supprimer in self.maMap.overlays
                {
                    //tester si la couche courante est celle qui
                    //affiche les trajets(les polylines) et
                    //la supprimer
                    if couche_a_supprimer is MKPolyline
                    {
                        maMap.remove(couche_a_supprimer)
                    }
            }
        
        //vider le tableau d'instuctions
        tbInstructions = [] 
    }
    
    
   
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        let destination = segue.destination
        
        if destination is InstuctionViewController
        {
        //récupérer l'écran de destination
        let destination = segue.destination as! InstuctionViewController
        
        destination.tbInformations = tbInstructions
        }
        
 
    }


}
